File: OCaml/hanoi.ml
open Stack

let hanoi n =
  let s1 = Stack.create () and s2 = Stack.create () and s3 = Stack.create () in
  let rec init n =
    if n <= 0 then ()
    else (
      push n s1;
      init (n - 1))
  in
  let print_stack s = Stack.iter (Printf.printf "%d # ") s in
  let print_state () =
    print_string "s1:";
    print_stack s1;
    print_string "\ns2:";
    print_stack s2;
    print_string "\ns3:\n";
    print_stack s3
  in
  let rec solve n =
    print_state ();
    if n = 0 then print_state ()
    else
      let p = pop s1 in
      push p s2;
      solve (n - 1);
      push 2 s3
  in
  init n;
  solve n

File: Java/Hanoi.java

/**
 * @author Edwin Ansari
 * The Hanoi problem recursive solution
 */
import java.util.Stack;

public class Hanoi {

    public static Stack<Integer> a1 = new Stack<>(), a2 = new Stack<>(), a3 = new Stack<>();
    public static int move = 0;

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        for (int i = n; i >= 1; i--) {
            a1.push(i);
        }
        printState();
        hanoi(n, a1, a2, a3);
    }

    public static void printState() {
        System.out.printf("MOVE#%d\ns1:%s\ns2:%s\ns3:%s\n", move++, a1, a2, a3);
    }

    public static void hanoi(int n, Stack<Integer> s1, Stack<Integer> s2, Stack<Integer> s3) {
        if (n <= 0) {
            return;
        } else if (n == 1) {
            s3.push(s1.pop());
            printState();
        } else {
            hanoi(n - 1, s1, s3, s2);
            hanoi(1, s1, s2, s3);
            hanoi(n - 1, s2, s1, s3);
        }
    }
}